platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsToolingApiModelQueryIntegrationTest.groovy

```diff
@@ -114,26 +114,73 @@ class IsolatedProjectsToolingApiModelQueryIntegrationTest extends AbstractIsolat
             }
         """
 
-        when:
+        when: "requesting a model together with running a task"
         withIsolatedProjects()
         fetchModel(SomeToolingModel, ":dummyTask")
 
-        then:
+        then: "only relevant projects are configured, while work graph and the model are stored in cache"
         fixture.assertModelStored {
+            runsTasks = true
             // TODO:isolated desired behavior
-//            projectsConfigured(":buildSrc", ":")
+//            projectsConfigured(":buildSrc", ":") // Note :a and :b were not configured
             projectsConfigured(":buildSrc", ":", ":a", ":b")
             modelsCreated(":")
         }
         outputContains("Configuration of dummyTask")
         outputContains("Execution of dummyTask")
 
-        when:
+        when: "repeating the request"
         withIsolatedProjects()
         fetchModel(SomeToolingModel, ":dummyTask")
 
-        then:
-        fixture.assertModelLoaded()
+        then: "no projects are configured, work graph and the model are loaded, tasks are executed before the model is returned"
+        fixture.assertModelLoaded {
+            runsTasks = true
+        }
+        outputDoesNotContain("Configuration of dummyTask")
+        outputDoesNotContain("creating model")
+        outputContains("Execution of dummyTask")
+    }
+
+    def "can cache models with tasks using internal option"() {
+        given:
+        withSomeToolingModelBuilderPluginInBuildSrc()
+        settingsFile << """
+            include("a")
+            include("b")
+        """
+        buildFile << """
+            plugins.apply(my.MyPlugin)
+
+            tasks.register("dummyTask") {
+                println("Configuration of dummyTask")
+                doLast {
+                    println("Execution of dummyTask")
+                }
+            }
+        """
+
+        when: "requesting a model together with running a task"
+        withIsolatedProjects("-Dorg.gradle.internal.isolated-projects.configure-on-demand.tasks=true")
+        fetchModel(SomeToolingModel, ":dummyTask")
+
+        then: "only relevant projects are configured, while work graph and the model are stored in cache"
+        fixture.assertModelStored {
+            runsTasks = true
+            projectsConfigured(":buildSrc", ":") // Note :a and :b were not configured
+            modelsCreated(":")
+        }
+        outputContains("Configuration of dummyTask")
+        outputContains("Execution of dummyTask")
+
+        when: "repeating the request"
+        withIsolatedProjects()
```

Comment:
🤔 do you intentionally omit the flag? Guess we'll revisit this if we make the flag value a CC input as the test will start to fail.

Reply:
I noticed that the flag is not part of the CC input, and this triggered the [question](https://gradle.slack.com/archives/C05Q92AUM8D/p1731657607513829) in Slack.

I think we do need to make a decision on whether these flags should be part of the input. There is no explicit test that captures that today. But let's do this in one of the follow-ups.

I'll repeat the flag in this test, because it's the intention, even if it currently doesn't have an effect.